### ISSUE TITLE
Zoom/Layout Refactor Design Document

### ISSUE THREAD

**Issue by redweasel** [xournalpp/xournalpp#3021]
In this ticket I will try to discuss most important aspects for a new zoom and layout system.
Feel free to join the discussion below.

## Goals
- **Must have**:
  - Page Layouts with 1-8 columns (skip first option for pair pages)
  - Support for different sized pages (to import old files at least)
  - Support for infinite page documents (as this kept users waiting for a long time now)
  - Stable Zoom/Scroll with Hand Tool, Mouse Wheel and Touch input
  - Button to Zoom current page to full width (which needs to be calculated for infinite pages) (once! not like currently)
- **Nice to have**:
  - Support for embedded pages in infinite pages
  - Rotation support with touch input.
  - Arrange Tool for page layout e.g. to plan out presentations on infinite pages.
  - Preferences for all buttons, modifiers and actions or at the very least an internal structure where the buttons are configured in one place so it is extendible to be used for preferences.
  - Lua api to make operators for page layout changes and zoom changes.
- **Must not have**:
  - Ability to interact with canvas while zooming (important to avoid this since this way drawing can always happen in screen space until the stroke is finished and put on the canvas/page)

## Discussion
There is (obviously) a layout system for pages in place, which works in most cases. It however for whatever reason... has constant pixel distances between the pages, which means that the entire layout needs to be calculated as soon as the zoom level changes. This is obviously no good, so the new layout system will only use Millimeters or Centimeters as unit which is in page space (as opposed to being in screen space). No discussion here really as this is pretty clear so far.

Now I want to dwell a little on zoom precision. There were zoom precision issues before in xournalpp, but they were not due to the double precision, but because pixel units (integer) where used. Double precision is relative to the value. It is 16 digits precise. To have a smooth feeling zoom we need at least pixel precision, which usually means around ~0.02cm. Lets go and support even high dpi screens, so say we want 0.001cm precision. Remember 0.001cm might be more than one pixel if zoomed very close up as I am talking about the translation (scroll) in page space in the following. But say 10^-4cm is our goal, then we can still have a document 10^(16-4)=10^12cm long, which is when considering a page is ~30cm long far over a billion pages until we reach our precision limit. So I think it is clear, that double precision really covers all our bases here. Now lets quickly take a look at float precision, cause float operations are faster and more compact in memory and on disk. Float has a precision of ~7digits which means that with 10^-4cm as precision goal, we can have a document 10^(7-4)/30=33.3 pages long until we exceed the precision limit. That is obviously not enough. But strokes for example are never bigger than a page, so maybe strokes should be computed and saved as floats (don't actually know how it is done currently, but am assuming it's double everywhere).

Back to actual zoom. Luckily double is enough precision, so we can build a very simple system for zooming. All we need is a single 3x3 (double-precision) matrix and maybe a cached inverse of it. It's common place to use 3x3 matrices for 2d transforms as you can use scew in 3D to make a 2D affine transform (meaning there can be translation). All events happen in screen space (which would be the space of the actual view into the pages, not the window containing the view) and then get processed there or get transferred to page space using matrix multiplication.

## Implementation (Workplan)
In this part I will propose a possible workplan to get the new zoom system.

**Part I**:
This first part can be done on master without affecting versioning:

- [ ] Include or write a vector and matrix library.
  - there are already a lot of components floating around in xournalpp, but they are all kind of disconnected, what is needed is a self contained package of all the math xournalpp needs for zoom (mat2, mat3, vec2), but also for rendering (rect, bvhtree) and for strokes with pressure (vec2, vec3, mat3). We should definitly move away from calling vectors `Point`.
- [ ] make all page distances in `LayoutHandler` zoom independent (use page space) and make updates to the page layout sparse (only when really needed).

**Part II**:
The following part of changes should be done collaboratively on a new official branch after the first part is done, because it can break files and may be unusable for some time.

- [ ] Remove `ZoomHandler` and `ScrollHandler` and replace it with a `ViewHandler` which uses the matrix to zoom and scroll and also has the ability to rotate build it.
- [ ] Add *usable* rotation ability to TouchInputHandler and clean up the code there along with it.
- [ ] Make all stroke drawings and similar element operators use screen space and copy on finish to the page they are on (or to the infinite page).
- [ ] Make strokes which are drawn over multiple pages split up and clip correctly
- [ ] Add structure for configurable key bindings and use that in all input handlers and gui shortcuts
  - [ ] Add a preferences section for keybindings
- [ ] Add BVH Tree for all elements on all pages to compute visible parts and clicked strokes when infinite pages come around (pages can be their own BVH Nodes in the global tree)
- [ ] Ininite page support
  - [ ] Update the LayoutHandler to support documents with manually arranged pages.
  - [ ] Add `control/tools/ArrangeHandler` tool with toggleable auto-rearrange option to move and resize pages with a hand like tool for pdf export and presentation target frames.
  - [ ] For infinite page documents: Make the background an infinite page with pages only being the windows for pdf export, printing and presentations. All data should be saved on the infinite page and selected for pdf export or zoomed to for presentations.
- [ ] https://github.com/xournalpp/xournalpp/pull/2729 (spline strokes) should go in the same time as this branch merges into master to avoid breaking files twice, so It would be good to commit it on the branch first.
  - [ ] make strokes use floats for their data in local stroke space and use a double position for their page space position.

If someone wants to tackle one of the tasks of the second part, it would be good to create a design task before starting to code, so the exact design can be worked out, before spending hours coding.

On all the points we should always strive for simplicity (a matrix multiplication is simpler than a matrix multiplication and an addition of a translation vector, one calculation which is valid for all cases is simpler then an if statement which handles each case individually, etc.)

**Further Discussion**:
There needs to be some more discussion on the following topics in separate sub tasks:
- (???) Infinite Pages Design
- (???) Preferences UI for Key Bindings
- (???) New Stroke Design with https://github.com/xournalpp/xournalpp/pull/2729 coming up

The list of things to do will get a little longer as soon as we get there I bet. If there is anything you want to add or change, feel free to comment and propose it. This really needs to be a community effort.

**Comment by Esokrates** [comment]
Do I understand right, that this is a subtask of https://github.com/xournalpp/xournalpp/issues/1795?
Don't know if it related, but is https://github.com/xournalpp/xournalpp/issues/1971 relevant here?

**Comment by redweasel** [comment]
I just tried to come up with some of my own thoughts of where to go with https://github.com/xournalpp/xournalpp/issues/1795
I especially wanted to include the infinity pages and the massive simplification of the ZoomControl in my design.

I dont think https://github.com/xournalpp/xournalpp/issues/1971 is relevant here, as this is about the widget to draw the viewport.
https://github.com/xournalpp/xournalpp/issues/1433 is relevant though, although it's really short and @lehmanju was not active in xournalpp for a while.
